src/libmongoc/src/mongoc/mongoc-scram.c

```diff
@@ -1052,93 +1051,185 @@ _mongoc_sasl_prep_impl (const char *name,
+   // requirements for bidirectional strings.  If the string does not
+   // satisfy the requirements for bidirectional strings, return an
+   // error.
+
+   // note: bidi stands for directional (text). Most characters are displayed
+   // left to right but some are displayed right to left. The requirements are
+   // as follows:
+   // 1. If a string contains any RandALCat character, it can't contain an LCat
+   // character
+   // 2. If it contains an RandALCat character, there must be an RandALCat
+   // character at the beginning and the end of the string (does not have to be
+   // the same character)
+   bool contains_LCat = false;
+   bool contains_RandALCar = false;
+
+
+   for (size_t i = 0; i < num_chars; ++i) {
+      if (_mongoc_utf8_code_point_is_in_table (utf8_codepoints[i],
+                                               LCat_bidi_ranges,
+                                               sizeof (LCat_bidi_ranges) /
+                                                  sizeof (uint32_t))) {
+         contains_LCat = true;
+         if (contains_RandALCar)
+            break;
+      }
+      if (_mongoc_utf8_code_point_is_in_table (utf8_codepoints[i],
+                                               RandALCat_bidi_ranges,
+                                               sizeof (RandALCat_bidi_ranges) /
+                                                  sizeof (uint32_t)))
+         contains_RandALCar = true;
+   }
+
+   if (
+      // requirement 1
+      (contains_RandALCar && contains_LCat) ||
+      // requirement 2
+      (contains_RandALCar &&
+       (!_mongoc_utf8_code_point_is_in_table (utf8_codepoints[0],
+                                              RandALCat_bidi_ranges,
+                                              sizeof (RandALCat_bidi_ranges) /
+                                                 sizeof (uint32_t)) ||
+        !_mongoc_utf8_code_point_is_in_table (utf8_codepoints[num_chars - 1],
+                                              RandALCat_bidi_ranges,
+                                              sizeof (RandALCat_bidi_ranges) /
+                                                 sizeof (uint32_t))))) {
       bson_free (out_utf8);
-      bson_free (out_utf16);
-      SASL_PREP_ERR_RETURN ("could not convert %s back to UTF-8");
+      bson_free (utf8_pre_norm);
+      bson_free (utf8_codepoints);
+      SASL_PREP_ERR_RETURN ("%s does not meet bidirectional requirements");
    }
-   bson_free (out_utf16);
-   return out_utf8;
+
+   bson_free (utf8_pre_norm);
+   bson_free (utf8_codepoints);
+
+   return (char *) out_utf8;
 #undef SASL_PREP_ERR_RETURN
 }
 #endif
 
 char *
-_mongoc_sasl_prep (const char *in_utf8, int in_utf8_len, bson_error_t *err)
+_mongoc_sasl_prep (const char *in_utf8, bson_error_t *err)
 {
-   BSON_UNUSED (in_utf8_len);
-
 #ifdef MONGOC_ENABLE_ICU
-   return _mongoc_sasl_prep_impl ("password", in_utf8, in_utf8_len, err);
+   return _mongoc_sasl_prep_impl ("password", in_utf8, err);
```

Comment:
```suggestion
   if (_mongoc_sasl_prep_required (in_utf8)) {
      return _mongoc_sasl_prep_impl ("password", in_utf8, err);
   }
   return bson_strdup (in_utf8);
```

Suggest only running `_mongoc_sasl_prep_impl` if a password includes non-ASCII characters. I expect the majority of passwords are ASCII only characters and do not need to undergo SASLPrep.